hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/merge/SparkRecordMergingUtils.java

```diff
@@ -0,0 +1,192 @@
+import org.apache.hudi.common.util.collection.Pair;
+
+import org.apache.avro.Schema;
+import org.apache.spark.sql.HoodieInternalRowUtils;
+import org.apache.spark.sql.catalyst.InternalRow;
+import org.apache.spark.sql.catalyst.expressions.GenericInternalRow;
+import org.apache.spark.sql.types.StructField;
+import org.apache.spark.sql.types.StructType;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Util class to merge records that may contain partial updates.
+ * This can be plugged into any Spark {@link HoodieRecordMerger} implementation.
+ */
+public class SparkRecordMergingUtils {
+  private static final Map<Schema, Map<Integer, StructField>> FIELD_ID_TO_FIELD_MAPPING_CACHE = new ConcurrentHashMap<>();
+  private static final Map<Schema, Map<String, Integer>> FIELD_NAME_TO_ID_MAPPING_CACHE = new ConcurrentHashMap<>();
+  private static final Map<Pair<Schema, Schema>,
+      Pair<Map<Integer, StructField>, Pair<StructType, Schema>>> MERGED_SCHEMA_CACHE = new ConcurrentHashMap<>();
+
+  /**
+   * Merges records which can contain partial updates.
+   *
+   * @param older     Older {@link HoodieSparkRecord}.
+   * @param oldSchema Old schema.
+   * @param newer     Newer {@link HoodieSparkRecord}.
+   * @param newSchema New schema.
+   * @param props     Configuration in {@link TypedProperties}.
+   * @return The merged record.
+   */
+  public static Pair<HoodieRecord, Schema> mergeCompleteOrPartialRecords(HoodieSparkRecord older,
+                                                                         Schema oldSchema,
+                                                                         HoodieSparkRecord newer,
+                                                                         Schema newSchema,
+                                                                         TypedProperties props) {
+    Pair<Map<Integer, StructField>, Pair<StructType, Schema>> mappingSchemaPair =
+        getCachedMergedSchema(oldSchema, newSchema);
+    boolean isNewerPartial = isPartial(newSchema, mappingSchemaPair.getRight().getRight());
+    if (isNewerPartial) {
+      InternalRow oldRow = older.getData();
+      InternalRow newPartialRow = newer.getData();
+
+      Map<Integer, StructField> mergedIdToFieldMapping = mappingSchemaPair.getLeft();
+      Map<String, Integer> newPartialNameToIdMapping = getCachedFieldNameToIdMapping(newSchema);
+      List<Object> values = new ArrayList<>(mergedIdToFieldMapping.size());
+      for (int fieldId = 0; fieldId < mergedIdToFieldMapping.size(); fieldId++) {
+        StructField structField = mergedIdToFieldMapping.get(fieldId);
+        if (newPartialNameToIdMapping.containsKey(structField.name())) {
+          // pick from new
+          int ordInPartialUpdate = newPartialNameToIdMapping.get(structField.name());
```

Comment:
Using `get` then decides val != null is always more efficient.

Reply:
Fixed.